### PR TITLE
feat(seeding): make seeder configureable

### PR DESCRIPTION
**Pull request by Phil91**
## Description

Add configuration to disable specific seeded values in the keycloak seeder

## Why

To be able to seed only wanted values

## Issue

Refs: #1172

## Corresponding CD PR

https://github.com/eclipse-tractusx/portal-iam/pull/234

## Checklist

- [x] I have followed the [contributing guidelines](https://github.com/eclipse-tractusx/portal-assets/blob/main/docs/admin/Dev%20Process/How%20to%20contribute.md#commits-branches-and-pull-requests-guidelines)
- [x] I have performed [IP checks](https://eclipse-tractusx.github.io/docs/release/trg-7/trg-7-04#checking-libraries-using-the-eclipse-dash-license-tool) for added or updated 3rd party libraries
- [x] I have created and linked IP issues or requested their creation by a committer
- [x] I have performed a self-review of my own code
- [x] I have successfully tested my changes locally
- [x] I have added tests that prove my changes work
- [x] I have checked that new and existing tests pass locally with my changes
- [x] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/keycloak/Keycloak.Seeding/Models/ConfigurationKeys.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+{
+    public const string RolesConfigKey = "ROLES";
+    public const string LocalizationsConfigKey = "LOCALIZATIONS";
+    public const string UserProfileConfigKey = "USERPROFILE";
+    public const string ClientScopesConfigKey = "CLIENTSCOPES";
+    public const string ClientsConfigKey = "CLIENTS";
+    public const string IdentityProvidersConfigKey = "IDENTITYPROVIDERS";
+    public const string IdentityProviderMappersConfigKey = "IDENTITYPROVIDERMAPPERS";
+    public const string UsersConfigKey = "USERS";
+    // TODO (PS): Clarify how to define the identity providers which should be skipped
```
Comment (evegufy) [inline src/keycloak/Keycloak.Seeding/Models/ConfigurationKeys.cs:32]: see https://github.com/eclipse-tractusx/portal-iam/blob/main/import/realm-config/generic/catenax-central/CX-Central-users-0.json#L21

File: src/keycloak/Keycloak.Seeding/Models/ConfigurationKeys.cs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    public const string FederatedIdentitiesConfigKeys = "FEDERATEDIDENTITIES";
+    public const string ClientScopeMappersConfigKey = "CLIENTSCOPEMAPPERS";
+    public const string ProtocolMappersConfigKey = "PROTOCOLMAPPERS";
+    // TODO (PS): Clarify how to define the auth flows which should be skipped
+    public const string AuthenticationFlowsConfigKey = "AUTHENTICATIONFLOWS";
+    public const string ClientProtocolMapperConfigKey = "CLIENTPROTOCOLMAPPER";
+    public const string ClientRolesConfigKey = "CLIENTROLES";
+    public const string AuthenticationFlowExecutionConfigKey = "AUTHENTICATIONFLOWEXECUTION";
+    public const string AuthenticatorConfigConfigKey = "AUTHENTICATORCONFIG";
```
Comment (evegufy) [inline src/keycloak/Keycloak.Seeding/Models/ConfigurationKeys.cs:41]: use enum

File: src/keycloak/Keycloak.Seeding/appsettings.json [hunk 3]
```diff
@@ -36,7 +36,111 @@
...
+              "Key": "ClientRoles",
+              "Create": true,
+              "Update": true,
+              "Delete": true
+            },
+            {
+              "Key": "AuthenticatorConfig",
+              "Create": true,
+              "Update": true,
+              "Delete": true
```
Comment (evegufy) [inline src/keycloak/Keycloak.Seeding/appsettings.json:140]: Please remove values that don't make sense (e.g. userProfiles) and document accordingly.

**Review by evegufy — CHANGES_REQUESTED** [review]

File: src/keycloak/Keycloak.Seeding/Models/ConfigurationKeys.cs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ********************************************************************************/
+
+namespace Org.Eclipse.TractusX.Portal.Backend.Keycloak.Seeding.Models;
+
+public enum ConfigurationKeys
```
Comment (ntruchsess) [inline src/keycloak/Keycloak.Seeding/Models/ConfigurationKeys.cs:22]: I'd not use the plural in enum-types as the parameter that is being passed to a method is only a single value
```suggestion
public enum ConfigurationKey
```

File: src/keycloak/Keycloak.Seeding/Models/KeycloakRealmSettings.cs [hunk 5]
```diff
@@ -32,6 +32,11 @@ public class KeycloakRealmSettings
     [Required]
     [DistinctValues]
     public IEnumerable<string> DataPaths { get; set; } = null!;
+    public bool Create { get; set; }
+    public bool Update { get; set; }
+    public bool Delete { get; set; }
+    [DistinctValues]
+    public IEnumerable<SeederConfiguration>? SeederConfiguration { get; set; }
```
Comment (ntruchsess) [inline src/keycloak/Keycloak.Seeding/Models/KeycloakRealmSettings.cs:39]:
```suggestion
    public IEnumerable<SeederConfiguration>? SeederConfigurations { get; set; }
```

File: src/keycloak/Keycloak.Seeding/BusinessLogic/ClientScopesUpdater.cs [hunk 6]
```diff
@@ -22,37 +22,42 @@
...
+    private static async Task CheckAndExecute(ConfigurationKeys configKey, ModificationType modificationType, KeycloakClient keycloak, string realm, IEnumerable<ClientScope> clientScopes, IEnumerable<ClientScopeModel> seedClientScopes, KeycloakRealmSettings seederConfig, CancellationToken cancellationToken, Func<KeycloakClient, string, IEnumerable<ClientScope>, IEnumerable<ClientScopeModel>, KeycloakRealmSettings, CancellationToken, Task> executeLogic)
     {
-        foreach (var deleteScope in clientScopes.ExceptBy(seedClientScopes.Select(x => x.Name), x => x.Name))
+        if (!seederConfig.ModificationAllowed(configKey, modificationType))
+        {
+            return;
+        }
+
+        await executeLogic(keycloak, realm, clientScopes, seedClientScopes, seederConfig, cancellationToken).ConfigureAwait(ConfigureAwaitOptions.None);
+    }
```
Comment (ntruchsess) [inline src/keycloak/Keycloak.Seeding/BusinessLogic/ClientScopesUpdater.cs:55]:
```suggestion
private static Task CheckAndExecute(ConfigurationKey configKey, ModificationType modificationType, KeycloakClient keycloak, string realm, IEnumerable<ClientScope> clientScopes, IEnumerable<ClientScopeModel> seedClientScopes, KeycloakRealmSettings seederConfig, CancellationToken cancellationToken, Func<KeycloakClient, string, IEnumerable<ClientScope>, IEnumerable<ClientScopeModel>, KeycloakRealmSettings, CancellationToken, Task> executeLogic) =>
        seederConfig.ModificationAllowed(configKey, modificationType)
            ? executeLogic(keycloak, realm, clientScopes, seedClientScopes, seederConfig, cancellationToken)
            : Task.CompletedTask;
```
(there are more occurences of this pattern that should be adjusted the same way)

**Review by ntruchsess — CHANGES_REQUESTED** [review]

**Review by evegufy — CHANGES_REQUESTED** [review]
Since https://github.com/eclipse-tractusx/portal-backend/pull/1174/commits/26ac07f279aab53a09f77ab3a0c1fd29eb40259d, [this configuration ](https://github.com/eclipse-tractusx/portal-iam/pull/234/files#diff-ee300b45d719481af34a480f94ace430d25f73c634f2591e9c1ba74b32e2cba1R193-R240)doesn't work anymore, for instance the cx-admin role isn't imported anymore, it works with the commit before https://github.com/eclipse-tractusx/portal-backend/pull/1174/commits/210bfb3a5473130e1b0661a61f93ca17f3366617

**Review by evegufy — APPROVED** [review]
I tested now everything successfully 👍

File: src/keycloak/Keycloak.Seeding/BusinessLogic/KeecloakSeeder.cs [hunk 7]
```diff
@@ -44,17 +46,27 @@ public async Task Seed(CancellationToken cancellationToken)
...
+
+    private async Task CheckAndExecuteUpdater(ConfigurationKeys configKey, string instanceName, Func<string, CancellationToken, Task> updaterExecution, CancellationToken cancellationToken)
+    {
+        if (!seedDataHandler.Configuration.IsModificationAllowed(configKey))
+        {
+            return;
+        }
+
+        await updaterExecution(instanceName, cancellationToken).ConfigureAwait(false);
+    }
```
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/keycloak/Keycloak.Seeding/BusinessLogic/KeecloakSeeder.cs]: implemented

File: src/keycloak/Keycloak.Seeding/Models/KeycloakRealmSettings.cs (hunk 5 above)
Comment (ntruchsess) [reply src/keycloak/Keycloak.Seeding/Models/KeycloakRealmSettings.cs]: implemented

File: src/keycloak/Keycloak.Seeding/BusinessLogic/ClientScopesUpdater.cs (hunk 6 above)
Comment (ntruchsess) [reply src/keycloak/Keycloak.Seeding/BusinessLogic/ClientScopesUpdater.cs]: implemented

**Review by ntruchsess — APPROVED** [review]

**Review by ntruchsess — APPROVED** [review]

**Review by ntruchsess — APPROVED** [review]

**Review by ntruchsess — APPROVED** [review]

**Review by ntruchsess — APPROVED** [review]

**Review by evegufy — APPROVED** [review]
tested again, looks good 👍

**Pull request merged**